### ISSUE TITLE
Custom Fields improvements

### ISSUE THREAD

**Issue by gabriele-v** [moneymanagerex/moneymanagerex#1585]
New *Custom Fields Manager* implementation is working under wxgtk:
- [x] It has "Personalize custom fields | Transaction" title - I don't understand transaction word here
- [x] `OK` and `Cancel` buttons should be replaced with `Close` button - there is no way to accept or not changes made
- [x] First field has *Description* name - is this description (string describing usage etc.) really or name (short id)?
- [x] *Default* field can has value not matching field type - this should be validated
- [x] *RegEx* can be used to validate numbers too
- [x] *Default* is not reactivated after type change
- [x] *Choices* accept duplicated entries like "a;a;a;a" (?)
- [x] *Massive Content Update* - unclear function description, user is asked questions without meaning steps executed
- [x] *Description* accepts existing predefined column names (like *Payee*) and duplicates (i.e. two custom fields with the same *Description*)
- [x] Add/Edit/Remove actions should be moved to buttons in the same line as `OK` from gear button - these actions are main functionalities of this dialog
- [x] Time type is useless IMHO - there should be *Date* and *Date&Time* types. Please imagine multiple transactions from last year with "9:00" or "12:00" value - how to sort them? How to build filter rule? Field allowing storing morning/evening/night/... info is useless for me.
- [x] "true" value entered ad *Default* for boolean is not working

Transaction dialog test results:
- [x] RegEx validation is not working
- [x] multiple with duplicates results in strange behavior
- [x] there is arbitrary precision forced for decimals (2 digits)
- [x] Time custom field is broken - I can set only hour part but this changes transaction date (!)
- [x] Date type custom field with no default value defined is rendered with value "1900-01-01" when empty. We must force some default value (like "now") or render this field as empty and allow to store NULL value.
    Now new transaction with date custom field not edited stores "1900-01-01" value in database file.

Others features needed:
- [ ] display them in transaction panel, filter and search using their values and see them in reports
- [ ] support for importing/exporting values from/to csv/qif files to custom fields
- [ ] controls (like customised transaction grid, predefined filter presets, reports...) must validate custom field used still exists and have correct type and not crash after user delete/edit custom field definition
- [ ] ability to set custom field values in recurring transaction templates and confirm/modify while entering recurring transaction (with custom field existence at this time)

**Comment by vomikan** [comment]
> It has "Personalize custom fields | Transaction" title - I don't understand transaction word here

It mean that this custom fields for Transaction dialog only. (not Account dialog, for example, but this not released yet).

**Comment by vomikan** [comment]
> First field has Description name

Here may be any user defined string.

**Comment by gabriele-v** [comment]
Could be renamed to "Name" maybe it's more clear

**Comment by slodki** [comment]
Name - used and field identifier, should be different then column names used in MMEX already (reserved)
Tooltip description - longer text describing field use showing as tooltip

**Comment by vomikan** [comment]
Date type custom field with no default value defined is rendered with value "1900-01-01" when empty.
We must force some default value (like "now") or render this field as empty and allow to store NULL value.
Now new transaction with date custom field not edited stores "1900-01-01" value in database file.

**Comment by slodki** [comment]
and finally #1298

**Comment by slodki** [comment]
It's rendered as below now:
![mmex1](https://user-images.githubusercontent.com/3155552/39515350-1da34a9a-4dfa-11e8-93e9-e4f7abdc9b29.png)

I don't like red color here. Maybe something like this will be better?
![mmex2](https://user-images.githubusercontent.com/3155552/39515356-2110e39a-4dfa-11e8-8af9-d348f931e9fe.png)

**Comment by vomikan** [comment]
What about to show custom field details in notes fields (yellow marked).

![image](https://user-images.githubusercontent.com/6836805/39705325-bafa08fe-5216-11e8-92d8-75d16441876b.png)

**Comment by slodki** [comment]
Will it be possible to remember custom column displayed or not in this panel for each account separately?
I want to display cust_column_A and cust_column_B for checking accounts but only cust_column_C for cash and credit cards.

**Comment by vomikan** [comment]
@gabriele-v  welcome.

**Comment by vomikan** [comment]
Suggestion for Reccuring Dialog design.

![image](https://user-images.githubusercontent.com/6836805/39840180-32f020ba-53e7-11e8-8883-ca6803531026.png)

![image](https://user-images.githubusercontent.com/6836805/39840187-3b318b24-53e7-11e8-9d4a-90ce5c0fbbdd.png)

**Comment by vomikan** [comment]
-[ ] display them in transaction panel
Regarding this issue. The columns number, name and positions are hardcoded. It's quite difficult to change this part of the code. I suggest to add 1-3 columns like UDFC01 ... UDFC03 into checking panel.
In New/Edit Custon Field dialog choice widget should be added "View on checking panel" [UDFC01,  UDFC03, UDFC03].
In case if some value assigned with UDFC01 ... UDFC03 it will be provided on Checking Panel.

**Comment by gabriele-v** [comment]
Yes, I've found the same issue too. Better to have up to UDFC05 and in edit custom field should be something more generic (considering they are not just for transaction), like "View on %s grid (if available)" where %s is the reftype.

**Comment by vomikan** [comment]
> "View on %s grid (if available)" where %s is the reftype.
This string is too long. It's OK for tooltip.
View on grid [UDFC01]

**Comment by vomikan** [comment]
>  RegEx validation is not working

I suggest to remove this functionality. It's too complex and Regular Expressions is difficult to understand for regular user.

Also Bool fields type is useless.

**Comment by gabriele-v** [comment]
it should be quite easy to implement as a check before saving and it's practically a standard everywhere for validation. If someone doesn't know it, there are lots of example on the web.

Bool it's a standard DB type, so why not? Do you have found some problems with it?

**Comment by vomikan** [comment]
No problems

**Comment by gabriele-v** [comment]
After last commit only these points are open, but they requires some planning (not for 1.3.5 I'd say):
- [ ]  display them in transaction panel, filter and search using their values and see them in reports
- [ ]  support for importing/exporting values from/to csv/qif files to custom fields
- [ ]  controls (like customised transaction grid, predefined filter presets, reports...) must validate custom field used still exists and have correct type and not crash after user delete/edit custom field definition
- [ ]  ability to set custom field values in recurring transaction templates and confirm/modify while entering recurring transaction (with custom field existence at this time)

**Comment by jkuj6** [comment]
> After last commit only these points are open, but they requires some planning (not for 1.3.5 I'd say):
> 
>     * [ ]   display them in transaction panel, filter and search using their values and see them in reports
> 
>     * [ ]   support for importing/exporting values from/to csv/qif files to custom fields
> 
>     * [ ]   controls (like customised transaction grid, predefined filter presets, reports...) must validate custom field used still exists and have correct type and not crash after user delete/edit custom field definition
> 
>     * [ ]   ability to set custom field values in recurring transaction templates and confirm/modify while entering recurring transaction (with custom field existence at this time)

@vomikan i been needing to filter by custom fields and currently there is no way. I found this where you state that you can implement this.  Would it be able to be implemented in the next milestone v1.5.13 ?

**Comment by vomikan** [comment]
@jkuj6 You can download the beta version of the app and check it out.

https://ci.appveyor.com/project/moneymanagerex/moneymanagerex/builds/42824427/job/9gyccofh1qpx2pdt/artifacts

**Comment by vomikan** [comment]
https://github.com/moneymanagerex/moneymanagerex/issues/1634

**Comment by jkuj6** [comment]
Im unable to check it out on Windows.  Will this be officially added/fixed for the Linux v1.5.14 release?

**Comment by DevelopmentAnything** [comment]
Unable to enter Korean in Tooltip item.
if type in Korean, it will be blank.
(any unicode characters)
@vomikan

**Comment by vomikan** [comment]
> Unable to enter Korean in Tooltip item.
> if type in Korean, it will be blank.
> (any unicode characters)
> @vomikan

Please register new bug request.

**Comment by DevelopmentAnything** [comment]
> 
https://github.com/moneymanagerex/moneymanagerex/issues/5875
like this? It's my first time.
